external/nyc311/tests.py

```diff
@@ -0,0 +1,69 @@
+from django.test import TestCase
+from unittest import mock
+import pandas as pd
+
+from .stat import get_311_statistics
+from .fetch import fetch_311_data, fetch_311_data_as_dataframe, get_311_data
+from .stub import fetch_311_data as fetch_311_data_stub
+from .models import NYC311Statistics, NYC311Complaint
+
+
+class FetchNYC311ConstraintsTests(TestCase):
+    @mock.patch("external.nyc311.fetch.fetch_311_data", fetch_311_data_stub)
+    def test_fetch_311_data_as_dataframe(self):
+        df = fetch_311_data_as_dataframe("11201")
+        self.assertTrue(isinstance(df, pd.DataFrame))
+
+    @mock.patch("external.nyc311.fetch.fetch_311_data", fetch_311_data_stub)
+    def test_get_311_data(self):
+        results = get_311_data("11201")
+        for complaint in results:
+            self.assertTrue(isinstance(complaint, NYC311Complaint))
+
+    @mock.patch("external.nyc311.fetch.Socrata")
+    def test_fetch_311_data(self, MockSocrata):  # noqa: N803
+        fetch_311_data("11201")
+
+        MockSocrata.assert_called_once()
+        client = MockSocrata()
+        client.get.assert_called_once()
+
+
+@mock.patch("external.nyc311.fetch.fetch_311_data", fetch_311_data_stub)
```

Comment:
What's the difference in annotating the class with @mock.patch as opposed to the methods? Does it just automatically annotate all of the class methods as @mock.patch?

Reply:
Yes I think class-wise @mock.patch will be applied to all of the test functions in the class 
But for the `FetchNYC311ConstraintsTests`, I specified mock.patch per each method because one method(`test_fetch_311_data `) uses different patch.

Reply:
got it! 👍